### DISCUSSION TITLE
Add new modification API for the extension where the event can't reach

### DISCUSSION THREAD

**Discussion by padaliyajay** [opencart/opencart#11770] [Ideas]
First of all, removing ocmod from opencart 4 is best decision because it was doing so much conflict between extensions. but as we know, an event can't reach everywhere so we may need one more parallel API which do the job which event can't do.

I have the idea which i was implemented in my own project mazaengine. let me explain in detail with example.

i am talking about a hook API. first let me show the implementation of it then i will show you how to use

Library file:
```
<?php
namespace Opencart\System\Library;

use Opencart\System\Engine\Action;

class hook
{
    private $hooks;
    private $registry;

    /**
     * Constructor
     */
    public function __construct($registry)
    {
        $this->registry = $registry;
        $this->db = $this->registry->get('db');
    }

    public function add(string $trigger, $data): void
    {
        $this->hooks[$trigger][] = $data;
    }

    /**
     * Call hook and fetch data
     * Use in php code of controller, model and library 
     * @param String $trigger
     * @param Array $data
     * @return Array
     */
    public function fetch(string $trigger, array $data = array()): array
    {
        $return = array();

        if (isset($this->hooks[$trigger])) {
            foreach ($this->hooks[$trigger] as $action) {
                if ($action instanceof Action) {
                    $cb_return = $action->execute($this->registry, $data);
                } else {
                    $cb_return = $action;
                }

                if ($cb_return instanceof \Exception) {
                    throw $cb_return;
                } elseif (is_array($cb_return)) {
                    $return = array_merge($return, $cb_return);
                } else if ($cb_return) {
                    $return[] = $cb_return;
                }
            }
        }

        return $return;
    }

    /**
     * Shortcode hook call for twig template
     * @param string $shortcode 
     * Ex. "product.12" this will return html code of product detail thumb (id:12)
     * Ex. "form_register_fields" this will return custom form fields in account register form
     * /
    public function shortcode(string $shortcode)
    {
        $data = explode('.', $shortcode);

        $trigger = array_shift($data);

        return implode('', $this->fetch($trigger, $data));
    }

    /**
     * Add hook is similar to add event
     * @param String $code hook code
     * @param String $trigger
     * @param String $action class method
     */
    public function addHook(string $code, string $trigger, string $action): void
    {
        $this->db->query("REPLACE INTO `" . DB_PREFIX . "hook` SET `code` = '" . $this->db->escape($code) . "', `trigger` = '" . $this->db->escape($trigger) . "', `action` = '" . $this->db->escape($action) . "'");
    }

    /**
     * Delete specific hook
     * @param String $trigger
     * @param String $action
     */
    public function deleteHook(string $trigger, string $action): void
    {
        $this->db->query("DELETE FROM `" . DB_PREFIX . "hook` WHERE `trigger` = '" . $this->db->escape($trigger) . "' AND `action` = '" . $this->db->escape($action) . "' LIMIT 1");
    }

    /**
     * disable specific hook
     * @param String $trigger
     * @param String $action
     */
    public function disableHook(string $trigger, string $action): void
    {
        $this->db->query("UPDATE `" . DB_PREFIX . "hook` SET `status` = 0 WHERE `trigger` = '" . $this->db->escape($trigger) . "' AND `action` = '" . $this->db->escape($action) . "' LIMIT 1");
    }

    /**
     * delete hooks by code
     * @param String $code
     */
    public function deleteHooks(string $code): void
    {
        $this->db->query("DELETE FROM `" . DB_PREFIX . "hook` WHERE `code` = '" . $this->db->escape($code) . "'");
    }
}
```

Startup:
```
<?php
namespace Opencart\Catalog\Controller\Startup;
class Hook extends \Opencart\System\Engine\Controller {
	public function index(): void {
                $this->hook = new Opencart\System\Library\Hook($this->registry);

                if ($this->config->has('action_hook')) {
                    foreach ($this->config->get('action_hook') as $trigger => $value) {
                        foreach ($value as $action) {
                            $this->hook->add($trigger, new \Opencart\System\Engine\Action($action));
                        }
                    }
                }
                
                $query = $this->db->query("SELECT * FROM `" . DB_PREFIX . "hook` WHERE status = 1");
                
                foreach($query->rows as $hook){
                       $this->hook->add($hook['trigger'], new \Opencart\System\Engine\Action($hook['action']));
                }
	}
}
```

Trigger: (Currently added only two trigger, but in real project we should implement many place where event does not reach)
1). Allow to modify where condition of sql query of getProducts
File: catalog/model/catalog/product.php

```
if (!empty($data['filter_manufacturer_id'])) {
	$sql .= " AND p.`manufacturer_id` = '" . (int)$data['filter_manufacturer_id'] . "'";
}

$this->hook->fetch('sql_products_where', [&$sql]);

$sql .= " GROUP BY p.product_id";
```

2). Allow to add new custom tab beside description and specifications  in product page
File: catalog/view/template/product/product.twig

```
<ul class="nav nav-tabs">
          <li class="nav-item"><a href="#tab-description" data-bs-toggle="tab" class="nav-link active">{{ tab_description }}</a</li>
          {% if attribute_groups %}
            <li class="nav-item"><a href="#tab-specification" data-bs-toggle="tab" class="nav-link">{{ tab_attribute }}</a></li>
          {% endif %}
          {% if review_status %}
            <li class="nav-item"><a href="#tab-review" data-bs-toggle="tab" class="nav-link">{{ tab_review }}</a></li>
          {% endif %}
          {% for item in hook.shortcode('product_tab.12') %}
            <li class="nav-item"><a href="#tab-custom-{{ item.id }}" data-bs-toggle="tab" class="nav-link">{{ item.name }}</a></li>
          {% endfor %}
</ul>
<!-- Do the same for tab content -->
```
Register for hook is same as register for event
```
$this->hook->addHook('my_custom_hook_idetify_code', 'sql_products_where', 'extension/example_extension/hooks/product|filter');
$this->hook->addHook('my_custom_hook_idetify_code', 'product_tab', 'extension/example_extension/hooks/product|tab');
// here third parameter is controller function location, which return hook output and accept hook parameter
```

This is not final implementation, i initially made in opencart 3 so required to change it as per opencart 4 coding standard. I want to take attention of @danielkerr on this, please let us know your thoughts

So now let's start discussion on this.

**Comment by osworx** [comment]
One a first view,it looks quite promising.

**Reply by padaliyajay** [reply]
Yes it is. Lets see what @danielkerr thoughts

**Comment by mhcwebdesign** [comment]
I am not quite sure why you call it hooks instead of events. All you basically want is more event triggers, especially for library classes. There will always be limitations on what can be modified.

Perhaps a more fundamental approach can solve the issue of more flexible modification capabilities in a better way.

I am thinking of a decorator design pattern (see e.g. https://refactoring.guru/design-patterns/decorator) in conjunction with a factory class (https://refactoring.guru/design-patterns/factory-method).

Also, I am planning to port our Override Engine (https://www.mhccorp.com/opencart/free-items/override-engine-oc3) to OpenCart 4.

**Reply by padaliyajay** [reply]
Hook name make it separate from traditional opencart event. i believe this style is more close to opencart coding standards without any big changes in core framework. and it is similar to current opencart event so it will easy to understand for everyone

You should add example of your idea so we can understand it. we just want whatever is best for more modification control

**Reply by punitkumar** [reply]
#1 Can this hook system is able to modify data return by getProducts() of cart.php library file?
#2 Can we attach a controller function to trigger as after event for add() of cart.php to perform some stuff after a product added into cart? 
For example: $this->cart->add(); , $this->cart->getProducts();

**Reply by padaliyajay** [reply]
@punitkumar 
this hook system is not implemented in an opencart yet.  but it can be possible if hook trigger added that place

**Reply by punitkumar** [reply]
@padaliyajay okay, right now I am not able to find way for my needs. The hook class code you shared is complete and can be used in Opencart to test the functionality you explained?

**Reply by padaliyajay** [reply]
@padaliyajay No, it is for demo only. i did not test it after changes into it

**Reply by punitkumar** [reply]
Okay, wish oc events provide more flexibility sooner or later.

**Comment by mhcwebdesign** [comment]
@punitkumar: Modifying methods from the Cart class, such as **getProducts** or **add**, would be easy with our planned Override Engine. It would merely be a matter of overriding these methods, they can still call the parent method if needed.

The problem with hooks and/or event is that you have to add pre-planned triggers in so many places, where most of them won't ever be used, and on the other hand, where some wanted trigger places could still be missing.

Thinking about it, even without our planned Override Engine, if we implemented a simpler decorator design pattern, you'd still gain more flexibility compared to events/hooks triggers.

**Reply by punitkumar** [reply]
@mhcwebdesign you are right, I believe if Opencart implement before/after events for library cart.php file in same fashion implemented for controller, model, view, etc it would be great. 

Override function might be good idea, if it does not create mess alike with poorly coded OCMOD.

**Reply by padaliyajay** [reply]
@punitkumar as of now, it is not implemented like that. so i recommend to create new cart class in your extension library which extend opencart cart class. then initialise that new cart class in startup file and assign to $this->cart

**Reply by padaliyajay** [reply]
@mhcwebdesign you are right about event/hook. it is cons of their

**Reply by punitkumar** [reply]
extend Opencart cart class is not success when two different modules extend opencart class class to implement the stuff. Module code that execute at last will be in effect, have look at this video https://www.loom.com/share/7e0c628ed42b45b2beca93999fa60340
